File: etherpad/src/etherpad/importexport/importexport.js
/**
 * Copyright 2009 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

jimport("java.io.File");
jimport("java.io.FileOutputStream");
jimport("java.lang.System.out.println");
jimport("java.io.ByteArrayInputStream");
jimport("java.io.ByteArrayOutputStream");
jimport("java.io.DataInputStream");
jimport("java.io.DataOutputStream");
jimport("net.appjet.common.sars.SarsClient");
jimport("com.etherpad.openofficeservice.OpenOfficeService");
jimport("com.etherpad.openofficeservice.UnsupportedFormatException");
jimport("com.etherpad.openofficeservice.TemporaryFailure");

import("etherpad.log");
import("etherpad.utils");
import("sync");
import("execution");
import("varz");
import("exceptionutils");
import("etherpad.pad.exporthtml");
import("etherpad.importexport.toMarkdown.toMarkdown");
import("etherpad.control.pad.pad_view_control");

function _log(obj) {
  log.custom("import-export", obj);
}

function onStartup() {
  execution.initTaskThreadPool("importexport", 1);
}

var formats = {
//  pdf: 'application/pdf',
//  doc: 'application/msword',
//  odt: 'application/vnd.oasis.opendocument.text',
  html: 'text/html; charset=utf-8',
  txt: 'text/plain; charset=utf-8',
  md : 'text/x-web-markdown',
  'native' : 'text/html; charset=utf-8',
}

function _createTempFile(bytes, type) {
        var f = File.createTempFile("ooconvert-", (type === null ? null : (type == "" ? "" : "."+type)));
	if (bytes) {
		var fos = new FileOutputStream(f);
		fos.write(bytes);
	}
	return f;
}

function _initConverterClient(convertServer) {
  if (convertServer) {
    var convertHost = convertServer.split(":")[0];
    var convertPort = Number(convertServer.split(":")[1]);
    if (! appjet.scopeCache.converter) {
      var converter = new SarsClient("ooffice-password", convertHost, convertPort);
      appjet.scopeCache.converter = converter;
      converter.setConnectTimeout(5000);
      converter.setReadTimeout(40000);
      appjet.scopeCache.converter.connect();
    }
    return appjet.scopeCache.converter;
  } else {
    return null;
  }
}

function _conversionSarsFailure() {
  delete appjet.scopeCache.converter;
}

function errorUnsupported(from) {
  return "Unsupported file type"+(from ? ": <strong>"+from+"</strong>." : ".")+" Etherpad can only import <strong>txt</strong>, <strong>html</strong>, <strong>rtf</strong>, <strong>doc</strong>, and <strong>docx</strong> files.";
}
var errorTemporary = "A temporary failure occurred; please try again later.";

function doSlowFileConversion(from, to, bytes, continuation) {
  var bytes = convertFileSlowly(from, to, bytes);
  //continuation.resume();
  return bytes;
}

function _convertOverNetwork(convertServer, from, to, bytes) {
  var c = _initConverterClient(convertServer);
  var reqBytes = new ByteArrayOutputStream();
  var req = new DataOutputStream(reqBytes);
  req.writeUTF(from);
  req.writeUTF(to);
  req.writeInt(bytes.length);
  req.write(bytes, 0, bytes.length);

  var retBtyes;
  try {
    retBytes = c.message(reqBytes.toByteArray());
  } catch (e) {
    if (e.javaException) {
      net.appjet.oui.exceptionlog.apply(e.javaException)
    }
    _conversionSarsFailure();
    return "A communications failure occurred; please try again later.";
  }

  if (retBytes.length == 0) {
    return "An unknown failure occurred; please try again later. (#5)";
  }
  var res = new DataInputStream(new ByteArrayInputStream(retBytes));
  var status = res.readInt();
  if (status == 0) { // success
    var len = res.readInt();
    var resBytes = java.lang.reflect.Array.newInstance(java.lang.Byte.TYPE, len);
    res.readFully(resBytes);
    return resBytes;
  } else if (status == 1) {
    return errorTemporary;
  } else if (status == 2) {
    var permFailureCode = res.readInt();
    if (permFailureCode == 0) {
      return "An unknown failure occurred. (#1)";
    } else if (permFailureCode == 1) {
      return errorUnsupported(from);
    }
  } else {
    return "An unknown failure occurred. (#2)";
  }
}

function convertFileSlowly(from, to, bytes) {
  var convertServer = appjet.config["etherpad.sofficeConversionServer"];
  if (convertServer) {
    return _convertOverNetwork(convertServer, from, to, bytes);
  }

  if (! utils.hasOffice()) {
    return "EtherPad is not configured to import or export formats other than <strong>txt</strong> and <strong>html</strong>. Please contact your system administrator for details.";
  }
  OpenOfficeService.setExecutable(appjet.config["etherpad.soffice"]);
  try {
    return OpenOfficeService.convertFile(from, to, bytes);
  } catch (e) {
    if (e.javaException instanceof TemporaryFailure) {
      return errorTemporary;
    } else if (e.javaException instanceof UnsupportedFormatException) {
      return errorUnsupported(from);
    } else {
      return "An unknown failure occurred. (#3)";
    }
  }
}

function _noteConversionAttempt() {
  varz.incrementInt("importexport-conversions-attempted");
}

function _noteConversionSuccess() {
  varz.incrementInt("importexport-conversions-successful");
}

function _noteConversionFailure() {
  varz.incrementInt("importexport-conversions-failed");
}

function _noteConversionTimeout() {
  varz.incrementInt("importexport-conversions-timeout");
}

function _noteConversionImpossible() {
  varz.incrementInt("importexport-conversions-impossible");
}

function precomputedConversionResult(from, to, bytes) {
  try {
    var retBytes = request.cache.conversionCallable.get(500, java.util.concurrent.TimeUnit.MILLISECONDS);
    var delay = Date.now() - request.cache.startTime;
    _log({type: "conversion-latency", from: from, to: to,
          numBytes: request.cache.conversionByteLength,
          delay: delay});
    varz.addToInt("importexport-total-conversion-millis", delay);
    if (typeof(retBytes) == 'string') {
      _log({type: "error", error: "conversion-failed", from: from, to: to,
            numBytes: request.cache.conversionByteLength,
            delay: delay});
      _noteConversionFailure();
    } else {
      _noteConversionSuccess();
    }
    return retBytes;
  } catch (e) {
    if (e.javaException instanceof java.util.concurrent.TimeoutException) {
      _noteConversionTimeout();
      request.cache.conversionCallable.cancel(false);
      _log({type: "error", error: "conversion-failed", from: from, to: to,
            numBytes: request.cache.conversionByteLength,
            delay: -1});
      return "Conversion timed out. Please try again later.";
    }
    _log({type: "error", error: "conversion-failed", from: from, to: to,
          numBytes: request.cache.conversionByteLength,
          trace: exceptionutils.getStackTracePlain(e)});
    _noteConversionFailure();
    return "An unknown failure occurred. (#4)";
  }
}

function convertFile(from, to, bytes) {
//  if (request.cache.conversionCallable) {
//    return precomputedConversionResult(from, to, bytes);
//  }

  _noteConversionAttempt();
  if (from == to) {
    _noteConversionSuccess();
    return bytes;
  }
  if (from == "txt" && to == "html") {
    _noteConversionSuccess();
    throw 'not implemented!';
    /*return (new java.lang.String(utils.renderTemplateAsString('pad/exporthtml.ejs', {
      content: String(new java.lang.String(bytes, "UTF-8")).replace(/&/g, "&amp;").replace(/</g, "&lt;"),
      pre: true
    }))).getBytes("UTF-8");*/
  }

  return doSlowFileConversion(from, to, bytes);

/*  request.cache.conversionByteLength = bytes.length;
  request.cache.conversionCallable =
    execution.scheduleTask("importexport", "doSlowFileConversion", 0, [
      from, to, bytes, request.continuation
    ]);
  request.cache.startTime = Date.now();
  request.continuation.suspend(45000);
  _noteConversionImpossible();
  return "An unexpected error occurred."; // Shouldn't ever get here.*/
}


function exportPadContent(pad, revisionId, format) {
  var out;
  switch(format) {
    case "txt":
      out = exporthtml.getPadPlainText(pad, revisionId);
      break;
    case "md":
      //out = exporthtml.getPadMarkdown(pad, revisionId);
      var html = exporthtml.getPadHTML(pad, revisionId);
      out = toMarkdown(html);
      break;
    case "native":
      out = pad_view_control.getPadHTML(pad, revisionId);
      break;
    case "html":
      out = exporthtml.getPadHTMLDocument(pad, revisionId);
      break;
    default:
      throw new Error("Unsupported Format.")
  }
  return out;
}

function contentTypeForFormat(format) {
  return formats[format];
}

function formatFileExtension(format) {
  return "." + format;
}

File: infrastructure/com.etherpad.openofficeservice/importexport.scala
/**
 * Copyright 2009 Google Inc.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.etherpad.openofficeservice;

import net.appjet.common.sars.{SarsServer,SarsMessageHandler};

import java.io.{DataInputStream,DataOutputStream};
import java.io.{File,FileOutputStream,ByteArrayInputStream,ByteArrayOutputStream};

/* Libraries needed for OO.org Conversion */
import com.sun.star.bridge.{XBridge,XBridgeFactory};
import com.sun.star.beans.{PropertyValue,XPropertySet};
import com.sun.star.connection.{NoConnectException,XConnection,XConnector};
import com.sun.star.container.XNamed;
import com.sun.star.document.{XExporter,XFilter};
import com.sun.star.frame.{XComponentLoader,XStorable};
import com.sun.star.lang.{XComponent,XMultiComponentFactory};
import com.sun.star.uno.{UnoRuntime,XComponentContext};

class OOSException(m: String) extends RuntimeException(m);
class UnsupportedFormatException(format: String) extends OOSException("Unsupported format: "+format);
object TemporaryFailure extends OOSException("Temporary failure");

object OpenOfficeServerUtility {
  def checkServerAvailability(host: String, port: Int): Boolean = {
    // Assume the server is running; this is the responsibility of the user
    return true;
  }
  def runOpenOfficeServer(path: String, host: String, port: Int, timeout: Int, wait: Boolean) {
    // nothing
  }
}

class OpenOfficeFileConverter {
  var host: String = "localhost";
  var port: Int = 8100;

  def setOpenOfficeServerDetails(host: String, port: Int) {
    this.host = host;
    this.port = port;
  }
  
  def convertFile(src: File, dst: File, converter: String, extension: String): Boolean = {
    try {
      val fromFile: String = "file:///" + src.getAbsolutePath();
      val toFile: String = "file:///" + dst.getAbsolutePath();

      val cnx: String = "socket,host="+this.host+",port="+this.port+"";
      val xRemoteContext: XComponentContext  = com.sun.star.comp.helper.Bootstrap.createInitialComponentContext(null);
      val x: Object = xRemoteContext.getServiceManager().createInstanceWithContext("com.sun.star.connection.Connector", xRemoteContext);
      val xConnector: XConnector  = UnoRuntime.queryInterface(classOf[XConnector], x).asInstanceOf[XConnector];
      val connection: XConnection  = xConnector.connect(cnx);

      if(connection == null) {
        throw new OOSException("Connection failure");
      }
      val x2: Object = xRemoteContext.getServiceManager().createInstanceWithContext("com.sun.star.bridge.BridgeFactory", xRemoteContext);
      val xBridgeFactory: XBridgeFactory = UnoRuntime.queryInterface(classOf[XBridgeFactory], x2).asInstanceOf[XBridgeFactory];
      val xBridge: XBridge = xBridgeFactory.createBridge("", "urp", connection, null);
      val x3: Object = xBridge.getInstance("StarOffice.ServiceManager");
      if (x3 == null) {
        throw new OOSException("Failed to get bridge");
      }

      val xMultiComponentFactory: XMultiComponentFactory  = UnoRuntime.queryInterface(classOf[XMultiComponentFactory], x3).asInstanceOf[XMultiComponentFactory];
      val xProperySet: XPropertySet  = UnoRuntime.queryInterface(classOf[XPropertySet], xMultiComponentFactory).asInstanceOf[XPropertySet];
      val oDefaultContext: Object  = xProperySet.getPropertyValue("DefaultContext");
      val xComponentContext: XComponentContext = UnoRuntime.queryInterface(classOf[XComponentContext], oDefaultContext).asInstanceOf[XComponentContext];

      val desktopObj: Object  = xMultiComponentFactory.createInstanceWithContext("com.sun.star.frame.Desktop", xComponentContext);
      val xcomponentloader: XComponentLoader = UnoRuntime.queryInterface(classOf[XComponentLoader], desktopObj).asInstanceOf[XComponentLoader];

      if(xcomponentloader == null) {
        throw new OOSException("XComponent Loader could not be loaded");
      }

      val loadProps: Array[PropertyValue] = new Array[PropertyValue](2);
      loadProps(0) = new PropertyValue();
      loadProps(0).Name = "Hidden";
      loadProps(0).Value = boolean2Boolean(false);

      loadProps(1) = new PropertyValue();
      loadProps(1).Name = "UpdateDocMode";
      loadProps(1).Value = "1";

      val component: XComponent = xcomponentloader.loadComponentFromURL(fromFile,"_blank", 0, loadProps);

      if (component == null) {
                       throw new OOSException("Failed to load document");
               }

               val convProps: Array[PropertyValue] = new Array[PropertyValue](2);
      convProps(0) = new PropertyValue();
      convProps(0).Name = "FilterName";
      convProps(0).Value = converter;

      val xstorable: XStorable = UnoRuntime.queryInterface(classOf[XStorable],component).asInstanceOf[XStorable];
      if (xstorable == null) {
          throw new OOSException("Storable could not be loaded");
      }
      xstorable.storeToURL(toFile, convProps);
      component.dispose();
      return true;
    }
    catch {
      case e => {
           e.printStackTrace();
                 throw new OOSException("Unknown exception occurred: "+e.getMessage());
                 }
    }
  }
}

object OpenOfficeService {
  val formats = Map(
    "pdf" -> "writer_pdf_Export",
    "doc" -> "MS Word 97",
    "html" -> "HTML (StarWriter)",
    "odt" -> "writer8",
    //"html" -> "XHTML Writer File",
    "txt" -> "Text"
  );

  def createTempFile(bytes: Array[Byte], suffix: String) = {
    var f = File.createTempFile("ooconvert-", if (suffix == null) { null } else if (suffix == "") { "" } else { "."+suffix });
  	if (bytes != null) {
  		val fos = new FileOutputStream(f);
  		fos.write(bytes);		
  	}
  	f;
  }

  var soffice = "soffice";
  def setExecutable(exec: String) {
    soffice = exec;
  }

  var openOfficeServerHost: String = "localhost";
  var openOfficeServerPort: Int = 8100;

  def setOpenOfficeServer(host: String, port: Int) {
    openOfficeServerHost = host;
    openOfficeServerPort = port;
  }

  def convertFile(from: String, to: String, bytes: Array[Byte]): Array[Byte] = {
    if (from == to) {
      return bytes;
    }

  	val tempFile = createTempFile(bytes, from);
  	val outFile = createTempFile(null, to);

       /*
        Just hardcoding server and port here.
        If you intend to use an Openoffice.org instance on a network machine,
        do it at your risk.

        Just, remember to setOpenOfficeServer from etherpad/importexport/importexport.js,
        Also, remember that OO.org is reading and writing files over file:/// URI. So, make sure that
        you can access the files from network machine. Hint, NFS. Not Need for Speed game, you idiot,
        Network File System.

       */

  	if (! OpenOfficeServerUtility.checkServerAvailability(openOfficeServerHost, openOfficeServerPort)) {
  		try {
  			OpenOfficeServerUtility.runOpenOfficeServer(soffice, openOfficeServerHost, openOfficeServerPort, 20000, true);
  		} catch {
  		  case e: java.io.IOException => {
  		    e.printStackTrace();
  		    throw TemporaryFailure;
    		}
  		}
  	}
  	var converter = new OpenOfficeFileConverter();
  	converter.setOpenOfficeServerDetails(openOfficeServerHost, openOfficeServerPort);
  	var status = false;
  	try {
  		status = converter.convertFile(tempFile, outFile, formats(to), to);
  	} catch {
  	  case e => {
  	    e.printStackTrace();
  		  throw new OOSException("Unknown exception occurred: "+e.getMessage());
		  }
  	}
  	if (status == false) {
  	  throw new UnsupportedFormatException(from);
  	}
  	net.appjet.common.util.BetterFile.getFileBytes(outFile);
  }

  def main(args: Array[String]) {
    if (args.length > 0) {
      soffice = args(0);
      if (soffice.length == 0) {
        System.exit(1)
      }
    }
    
    // Query format:
    // from: String, to: String, count: Int, bytes: Array[Byte]
    // Response format:
    // status: Int, <data>
    //   status 0 (success) - <data>: count: Int, bytes: Array[Byte]
    //   status 1 (temporary failure) - <data>: <none>
    //   status 2 (permanent failure) - <data>: type: Int
    //               type - 0: unknown failure.
    //                    - 1: unsupported format
    val handler = new SarsMessageHandler {
      override def handle(b: Array[Byte]): Option[Array[Byte]] = {
        val is = new DataInputStream(new ByteArrayInputStream(b));
        val from = is.readUTF;
        val to = is.readUTF;
        val len = is.readInt;
        val bytes = new Array[Byte](len);
        is.readFully(bytes);
        var status = 0;
        var permfailuretype = 0;
        
        println("Converting "+from+" -> "+to+" ("+len+" bytes)");
        
        val output = try {
          convertFile(from, to, bytes);
        } catch {
          case TemporaryFailure => {
            status = 1;
            null;
          }
          case e: UnsupportedFormatException => {
            status = 2;
            permfailuretype = 1;
            null;
          }
          case e => {
            status = 2;
            permfailuretype = 0;
            e.printStackTrace();
            null;
          }
        }
        
        val retBytes = new ByteArrayOutputStream();
        val ret = new DataOutputStream(retBytes);
        if (status != 0) {
          ret.writeInt(status); // error
          status match {
            case 2 => {
              ret.writeInt(permfailuretype);
            }
            case _ => { }
          }
        } else {
          ret.writeInt(0); // success
          ret.writeInt(output.length);
          ret.write(output, 0, output.length);
        }
        Some(retBytes.toByteArray());
      }
    }
    
    val server = new SarsServer("ooffice-password", handler, None, 8101);
    server.start();
    println("Server running...");
    server.join();
    println("Server quitting...");
  }
}





